bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java

```diff
@@ -0,0 +1,244 @@
+            var dir = storage.resolve(sanitizedUrl);
+            Repository repo;
+            if (!Files.exists(dir)) {
+                log.info("Cloning " + hostedRepo.getName());
+                Files.createDirectories(dir);
+                repo = Repository.clone(hostedRepo.getUrl(), dir);
+            } else {
+                log.info("Found existing scratch directory for " + hostedRepo.getName());
+                repo = Repository.get(dir)
+                        .orElseThrow(() -> new RuntimeException("Repository in " + dir + " has vanished"));
+            }
+
+            repo.fetchAll();
+            var depsFile = repo.root().resolve(depsFileName);
+
+            List<Branch> ordered = orderedBranches(repo, depsFile);
+            log.info("Merge order " + ordered);
+            for (Branch branch : ordered) {
+                log.info("Processing branch " + branch + "...");
+                repo.checkout(branch);
+                Set<String> parents = new HashSet<>(
+                        Files.exists(depsFile) ? Files.readAllLines(depsFile) : List.of("master"));
+                List<String> failedMerges = new ArrayList<>();
+                boolean progress;
+                boolean failed;
+                do {
+                    // We need to attempt merge parents in any order that works. Keep merging
+                    // and pushing, until no further progress can be made.
+                    progress = false;
+                    failed = false;
+                    for (var parentsIt = parents.iterator(); parentsIt.hasNext();) {
+                        String parent = parentsIt.next();
+                        try {
+                            mergeIfAhead(repo, branch, parent);
+                            progress = true;
+                            parentsIt.remove(); // avoid doing pointless merges
+                        } catch(IOException e) {
+                            log.severe("Merge with " + parent + " failed. Reverting...");
+                            repo.abortMerge();
+                            failedMerges.add(branch + " <- " + parent);
+                            failed = true;
+                        }
+                    }
+                } while(progress && failed);
+
+                if (!failedMerges.isEmpty()) {
+                    throw new IOException("There were failed merges:\n" + failedMerges);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        log.info("Ending topobot run");
+    }
+
+    private List<Branch> orderedBranches(Repository repo, Path depsFile) throws IOException {
+        List<Edge> deps = new ArrayList<>();
+        for (Branch branch : branches) {
+            repo.checkout(branch);
+            if (Files.exists(depsFile)) {
+                Files.lines(depsFile)
+                        .forEach(dep -> deps.add(new Edge(new Branch(dep), branch)));
```

Comment:
May I suggest having a function for reading `depsFile` that return the name of the branches in the depsFile as a `List`. Something like:
```java
private List<Branch> dependencies(Path depsFile) {
    if (Files.exists(depsFile)) {
        // parse despFile
    } else {
        return List.of("master");
    }
}